vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+#include "arch_types.h"
+
+/* VPR header files go second*/
+#include "vpr_types.h"
+#include "rr_graph_fwd.h"
+
+class RRGraph {
+  public: /* Types */
+    /* Iterators used to create iterator-based loop for nodes/edges/switches/segments */
+    typedef vtr::vector<RRNodeId, RRNodeId>::const_iterator node_iterator;
+    typedef vtr::vector<RREdgeId, RREdgeId>::const_iterator edge_iterator;
+    typedef vtr::vector<RRSwitchId, RRSwitchId>::const_iterator switch_iterator;
+    typedef vtr::vector<RRSegmentId, RRSegmentId>::const_iterator segment_iterator;
+
+    /* Ranges used to create range-based loop for nodes/edges/switches/segments */
+    typedef vtr::Range<node_iterator> node_range;
+    typedef vtr::Range<edge_iterator> edge_range;
+    typedef vtr::Range<switch_iterator> switch_range;
+    typedef vtr::Range<segment_iterator> segment_range;
+
+  public: /* Accessors */
+    /* Aggregates: create range-based loops for nodes/edges/switches/segments
+     * To iterate over the nodes/edges/switches/segments in a RRGraph, 
+     *    using a range-based loop is suggested.
+     *  -----------------------------------------------------------------
+     *    Example: iterate over all the nodes
+     *      for (const RRNodeId& node : nodes()) {
+     *        // Do something with each node
+     *      }
+     *
+     *      for (const RREdgeId& edge : edges()) {
+     *        // Do something with each edge
+     *      }
+     *
+     *      for (const RRSwitchId& switch : switches()) {
+     *        // Do something with each switch
+     *      }
+     *
+     *      for (const RRSegmentId& segment : segments()) {
+     *        // Do something with each segment
+     *      }
+     */
+    node_range nodes() const;
+    edge_range edges() const;
+    switch_range switches() const;
+    segment_range segments() const;
+
+    /* Node-level attributes */
+    size_t node_index(const RRNodeId& node) const; /* TODO: deprecate this accessor as outside functions should use RRNodeId */
+
+    /* get the type of a RRGraph node : types of each node, can be channel wires (CHANX or CHANY) or 
+     *                                  logic block pins(OPIN or IPIN) or virtual nodes (SOURCE or SINK)
+     *                                  see t_rr_type definition for more details
+     */
+    t_rr_type node_type(const RRNodeId& node) const;
+
+    /* Get coordinate of a node. (xlow, xhigh, ylow, yhigh):
+     *   For OPIN/IPIN/SOURCE/SINK, xlow = xhigh and ylow = yhigh
+     *   FIXME: should confirm if this is still the case when a logic block has a height > 1
+     *   For CHANX/CHANY, (xlow, ylow) and (xhigh, yhigh) represent
+     *   where the routing segment starts and ends.
+     *   Note that our convention alway keeps
+     *   xlow <= xhigh and ylow <= yhigh
+     *   Therefore, (xlow, ylow) is a starting point for a CHANX/CHANY in INC_DIRECTION
+     *   (xhigh, yhigh) is a starting point for a CHANX/CHANY in INC_DIRECTION
```

Comment:
DEC_DIRECTION (for second line)

Reply:
Also say this is for unidirectional wires, where there is only a single drive point.